src/core/response/parameters.rs

```diff
@@ -47,13 +57,48 @@ impl TryFrom<Json> for VpToken {
     type Error = Error;
 
     fn try_from(value: Json) -> Result<Self, Self::Error> {
-        serde_json::from_value(value).map(Self).map_err(Into::into)
+        match value {
+            // NOTE: When parsing a Json string, it must be base64Url encoded,
+            // therefore the base64 encoded string is decoded for internal representation
+            // of the VP token.
+            Json::String(s) => Ok(Self::Single(BASE64_URL_SAFE_NO_PAD.decode(s)?)),
+            // NOTE: When the Json is an object, it must be a map.
+            Json::Object(map) => Ok(Self::SingleAsMap(map)),
+            Json::Array(arr) => {
+                let mut tokens = Vec::new();
+                for value in arr {
+                    tokens.push(Self::try_from(value)?);
+                }
+                Ok(Self::Many(tokens))
+            }
+            _ => Err(Error::msg("Invalid vp_token")),
+        }
     }
 }
 
-impl From<VpToken> for Json {
-    fn from(value: VpToken) -> Self {
-        value.0.into()
+impl TryFrom<VpToken> for Json {
+    type Error = Error;
+
+    fn try_from(value: VpToken) -> Result<Self, Self::Error> {
+        match value {
+            VpToken::Single(s) => Ok(serde_json::Value::String(BASE64_URL_SAFE_NO_PAD.encode(s))),
+            VpToken::SingleAsMap(map) => Ok(serde_json::Value::Object(map)),
+            VpToken::Many(tokens) => {
+                let mut arr: Vec<Json> = Vec::new();
+                for token in tokens {
+                    arr.push(token.try_into()?);
+                }
+                Ok(arr.into())
+            }
+        }
+    }
+}
```

Comment:
This can be rewritten as `From<VpToken> for Json`:
```suggestion
impl From<VpToken> for Json {
    fn from(value: VpToken) -> Self {
        match value {
            VpToken::Single(s) => serde_json::Value::String(BASE64_URL_SAFE_NO_PAD.encode(s)),
            VpToken::SingleAsMap(map) => serde_json::Value::Object(map),
            VpToken::Many(tokens) => Self::Array(tokens.into_iter().map(Self::from).collect()),
        }
    }
}
```